src/tools/rbd/action/Export.cc

```diff
@@ -116,34 +117,94 @@ static int do_export(librbd::Image& image, const char *path, bool no_progress)
   }
 
   utils::ProgressContext pc("Exporting image", no_progress);
-
   SimpleThrottle throttle(max_concurrent_ops, false);
   uint64_t period = image.get_stripe_count() * (1ull << info.order);
+
+  size_t file_size = 0;
+  if (with_snaps) {
+    // header
+    bufferlist bl;
+    bl.append(utils::RBD_IMAGE_BANNER);
+
+    uint64_t size = info.size;
+    ::encode(size, bl);
+
+    r = bl.write_fd(fd);
+    if (r < 0) {
+      pc.fail();
+      close(fd);
+
+      return r;
+    }
+    file_size += bl.length();
+    lseek(fd, file_size, SEEK_SET);
+  }
+
   for (uint64_t offset = 0; offset < info.size; offset += period) {
     if (throttle.pending_error()) {
       break;
     }
 
     uint64_t length = min(period, info.size - offset);
-    C_Export *ctx = new C_Export(throttle, image, offset, length, fd);
+    C_Export *ctx = new C_Export(throttle, image, file_size + offset, offset, length, fd);
     ctx->send();
 
-    pc.update_progress(offset, info.size);
+    pc.update_progress(offset, info.size * 2);
   }
 
+  file_size += info.size;
   r = throttle.wait_for_ret();
   if (!to_stdout) {
     if (r >= 0) {
-      r = ftruncate(fd, info.size);
+      r = ftruncate(fd, file_size);
+      r = lseek(fd, file_size, SEEK_SET);
     }
-    close(fd);
   }
 
-  if (r < 0) {
-    pc.fail();
-  } else {
-    pc.finish();
+  if (with_snaps) {
+    std::vector<librbd::snap_info_t> snaps;
+    int r = image.snap_list(snaps);
+    if (r < 0) {
+      pc.fail();
+      return r;
+    } else if (0 == snaps.size()) {
+      return 0;
+    } else {
+      // header for snapshots
+      bufferlist bl;
+      bl.append(utils::RBD_IMAGE_SNAPS_BANNER);
+
+      uint64_t snap_num = snaps.size() + 1;
+      ::encode(snap_num, bl);
+
+      r = bl.write_fd(fd);
+      if (r < 0) {
+	return r;
+      }
+
+      const char *last_snap = NULL;
+      for (size_t i = 0; i < snaps.size(); ++i) {
+	utils::snap_set(image, snaps[i].name.c_str());
+	r = utils::export_diff_fd(image, last_snap, snaps[i].name.c_str(), false, fd, true);
```

Comment:
Understood -- but my rationale is that if I have 10 snapshots on a 100G image, instead of having the progress bar smoothly progress, it will jump as each snapshot is completed. It would be nice to have the granular feedback.